Add TLS support for discovery backend

dhiltgen opened the pull request:
This leverages recent additions to libkv enabling client
authentication via TLS so the discovery back-end can be locked
down with mutual TLS.  Example usage:

```
docker daemon [other args] \
    --cluster-advertise 192.168.122.168:2376 \
    --cluster-store etcd://192.168.122.168:2379 \
    --cluster-store-opt kv.cacertfile=/path/to/ca.pem \
    --cluster-store-opt kv.certfile=/path/to/cert.pem \
    --cluster-store-opt kv.keyfile=/path/to/key.pem
```

Signed-off-by: Daniel Hiltgen daniel.hiltgen@docker.com

Commit:
Add TLS support for discovery backend

This leverages recent additions to libkv enabling client
authentication via TLS so the discovery back-end can be locked
down with mutual TLS.  Example usage:

    docker daemon [other args] \
        --cluster-advertise 192.168.122.168:2376 \
        --cluster-store etcd://192.168.122.168:2379 \
        --cluster-store-opt kv.cacertfile=/path/to/ca.pem \
        --cluster-store-opt kv.certfile=/path/to/cert.pem \
        --cluster-store-opt kv.keyfile=/path/to/key.pem

Signed-off-by: Daniel Hiltgen <daniel.hiltgen@docker.com>

calavera commented:
This is going to need tests and documentation :smile_cat:

dhiltgen commented:
Test and doc added.

abronan commented:
LGTM overall!

I wonder if we shouldn't add an `InsecureSkipVerify` argument to the discovery init so that there is a possibility to use self-signed certs for testing (I changed this for Consul in docker/libkv#70).

vdemeester commented:
design LGTM

abronan commented:
ping @aluzzardi

dhiltgen commented:
@abronan InsecureSkipVerify sounds useful.  Since libkv is vendored, we'll need to bump it after your change merges.  I'm wondering if it would make sense to do that as a second change after this goes in, since it sets up the framework to easily add additional options settings.

albers commented:
I will add this to bash completion once it is merged.

abronan commented:
@dhiltgen I'm good with the PR as it is and including the additional change in another PR once this gets merged in `libkv` :+1:

aluzzardi commented:
Maybe `--cluster-store-opt` since those options are only tied to the store, not the rest of clustering?

aluzzardi commented:
Also, to be consistent with `--storage-opt` and such, do we want to prefix options with the driver's name?

e.g. `--cluster-store-opt etcd.keyfile`?

dhiltgen commented:
I've switched to --cluster-store-opt etcd.keyfile style flags. Since the structure in libkv is currently generic, I didn't add any logic to validate the driver and flags - it's just hardcoded with etcd...  Once the consul change comes in, and we figure out how to do auth on other stores we can refine that logic in pkg/discovery/kv/kv.go

NathanMcCauley commented on pkg/discovery/discovery.go:
```diff
@@ -28,7 +28,7 @@ type Backend interface {
 
 	// Initialize the discovery with URIs, a heartbeat and a ttl.
-	Initialize(string, time.Duration, time.Duration) error
+	Initialize(string, time.Duration, time.Duration, map[string]string) error
```
new `map[string]string` param needs a comment describing its purpose, like the other params.

dhiltgen replied:
Will do.

NathanMcCauley commented on pkg/discovery/kv/kv.go:
```diff
@@ -63,9 +63,23 @@ func (s *Discovery) Initialize(uris string, heartbeat time.Duration, ttl time.Du
 	s.ttl = ttl
 	s.path = path.Join(s.prefix, discoveryPath)
 
+	var config *store.Config
```
Please see if you can use the TLSConfig package here: https://github.com/docker/docker/blob/master/pkg/tlsconfig/config.go

It does a good job of setting defaults (ciphersuite options, etc)

dhiltgen replied:
I don't think this will be possible with the current structure of libkv and etcd.

When ClientTLS settings are passed, calling through https://github.com/docker/libkv/blob/master/store/etcd/etcd.go#L63 we wind up at https://github.com/coreos/go-etcd/blob/master/etcd/client.go#L115 and then wiring up the TLS at https://github.com/coreos/go-etcd/blob/master/etcd/client.go#L194

However, If I *also* populate the libkv's TLS config option with the client defaults as you suggest (https://github.com/docker/libkv/blob/master/store/store.go#L43) then *after* the transport is established (calls above) libkv's etcd impl clobbers the TLS via https://github.com/docker/libkv/blob/master/store/etcd/etcd.go#L76 and https://github.com/docker/libkv/blob/master/store/etcd/etcd.go#L100  The end result is connections fail with something along these lines:

time="2015-10-01T17:03:00.776288957Z" level=error msg="Registering as \"192.168.122.223:12376\" in discovery failed: Put https://192.168.122.223:12379/v2/keys/docker/nodes/192.168.122.223:12376: x509: certificate signed by unknown authority"

dhiltgen replied:
@abronan not sure what you think, but this might be an area we could improve libkv's impl to try to union the TLS settings together instead of clobbering them, although that might get messy, so it seems like the right answer might be pushing settings all the way to the etcd go library.

abronan replied:
@dhiltgen So what's happening now is that I really delegate the work of initializing TLS to the etcd client. For Consul it's different, I have some control over what I give to the `HttpTransport`. So this is something that really is on the libkv side. I don't think this impacts this PR in some way.

dhiltgen replied:
I think I've got an idea on how I can address @NathanMcCauley's suggestion - I'll update the PR shortly once I have it implemented and tested.

jfrazelle replied:
was this adressed

jfrazelle replied:
ah looks like it was , thanks

dhiltgen replied:
Yes.  It's a little confusing because of my workaround for the bug in the existing/old libkv+etcd.  I have to set up both the ClientTLS as well as TLS, but can confirm that the TLS field is used, which I've wired up as @NathanMcCauley requested.

dhiltgen commented:
Updated to address @NathanMcCauley's comments.  While making the change, I found a glitch in the underlying libkv/etcd implementation.  My plan was to *only* set the TLS setting, but that resulted in it incorrectly using http instead of https.  By setting the ClientTLS as well, it correctly gets wired up for https, but the passed in TLS settings are the ones ultimately used, which are based on the tlsconfig package.

@aluzzardi while going through the code, I realized that this approach (passing in the TLS settings) makes this more generic so I changed the keys to "kv.xxx".  Both consul and etcd use it, and it should be possible to fix up libkv's back-end for zookeeper to use it in the future as well.

aluzzardi commented:
@dhiltgen I think it's impossible to get zk support, @abronan correct me if I'm wrong

dhiltgen commented:
Per https://cwiki.apache.org/confluence/display/ZOOKEEPER/ZooKeeper+SSL+User+Guide it looks like Zookeeper itself supports client connections via TLS, and x509 based auth.  Looking at the library libkv uses, I believe the zookeeper driver could be updated to use ConnectWithDialer (https://github.com/samuel/go-zookeeper/blob/master/zk/conn.go#L138) with a thin shim that uses crypto/tls/Dial to get the communication channel set up.

dhiltgen commented:
I just rebased, and https://github.com/docker/docker/pull/16740 was recently merged, which appears to make the tests in kv_test bogus.  I'm not sure yet how to fix my test logic so it actually exercises the relevant code, since libkv mock store is no longer usable.

dhiltgen commented:
For my purposes, I think I can replicate part of libkv/store/mock.go without requiring testify just for initialization and recycle that so I can exercise my code, but I still worry about the rest of the tests in kv_test.go not actually testing things any more...

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "discovery_tls" git@github.com:dhiltgen/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dhiltgen commented:
Once https://github.com/docker/libkv/pull/75 is merged and vendored in, I'll squash the second commit on this change.

If we decide to back out https://github.com/docker/docker/pull/16740 then all the extra Mock code in kv_test can be removed.  If we decide to keep that change, then we can potentially enhance that new Mock driver a little so it can be used for other test cases in kv_test.

dhiltgen commented:
I've added a change to this PR bumping the libkv and its dependencies.  @jfrazelle would love you to take a look at how I'm doing the vendoring tweaks in https://github.com/dhiltgen/docker/commit/8f9be3e67943a3f2561040192233795b6b5a56b7 to get the code happy.  An example of the type of imports that needed fixing is https://github.com/coreos/etcd/blob/master/client/client.go#L30

icecrime commented:
So, d4c504788bdf1118859a3a13e122b973c2cc36d5 LGTM, but the vendoring part is quite scary :fearful: I'm not sure if we can do anything better though, will wait for @abronan and @jfrazelle expertise on this.

jfrazelle commented on hack/vendor.sh:
```diff
@@ -25,13 +25,16 @@ clone git github.com/armon/go-metrics eb0af217e5e9747e41dd5303755356b62d28e3ec
 clone git github.com/samuel/go-zookeeper d0e0d8e11f318e000a8cc434616d69e329edc374
 clone git github.com/deckarep/golang-set ef32fa3046d9f249d399f98ebaf9be944430fd1d
-clone git github.com/coreos/go-etcd v2.0.0
+clone git github.com/coreos/etcd ba949ae2be9a2dee53e09c5c0ad365233a4a4711
```
i just i dont know

jfrazelle replied:
why cant we run etcd as a cluster like most people does it have to be embedded

jfrazelle replied:
wait i dont understand dont they have a client library or something

icecrime replied:
They do: it's the [`client/` subdirectory of the `coreos/etcd` repo](https://github.com/coreos/etcd/tree/master/client), so I believe this line is fine. However, we drag another big one with it (such as `github.com/ugorji/go` below).

Also, the fact that imports are rewritten in etcd required @dhiltgen to have this `fiximports` function.

icecrime replied:
Related: docker/libkv#75

jfrazelle commented:
also @albers you are awesome :)

dhiltgen commented:
@jfrazelle I'm definitely open to suggestions on different/better ways to make the etcd client play nice with our vendoring.  The old go lib for etcd that libkv was using is deprecated, and the bump of libkv fixes some other bugs for the networking folks.  @mavenugo can elaborate a bit more on that if you're curious.

icecrime commented:
After investigation, I don't see a better way:
- `coreos/go-etcd` is deprecated
- `coreos/etcd/client` is quite big, and requires another quite big dependency (`github.com/ugorji/go/codec`)

I won't evaluate the `fiximports` solution, but LGTM on the rest.

jfrazelle commented:
ping @tianon for vendoring changes we were just discussinng the impact all the changes from even 1.7 to 1.8 have on packagers..

dhiltgen commented:
Addressed @tianon's comments - vendor script cleanups and pinned etcd to v2.2.0

icecrime commented:
Just adding here because I see a lot of confusion.

First thing: remember that GitHub *lies* about line changes when the commits are too big for it to display. In this case, this is the true number:

```
cd392ebdae3e6a0dc76cc558bdf149aedfe6a864 13 files changed, 205 insertions(+), 21 deletions(-)
345cb1ab005276af2fc643a5261a02f7b10c0a3d 75 files changed, 44676 insertions(+), 2039 deletions(-)
```

That being said, the `etc/client` is indeed _just_ the client, but it has a ridiculously huge dependency on a thing called [`github.com/ugorji/go`](github.com/ugorji/go):

```
vendor/src/github.com/coreos/etcd/LICENSE                       |   202 ++
 vendor/src/github.com/coreos/etcd/client/README.md              |    94 +
 vendor/src/github.com/coreos/etcd/client/auth_role.go           |   235 ++
 vendor/src/github.com/coreos/etcd/client/auth_user.go           |   297 +++
 vendor/src/github.com/coreos/etcd/client/cancelreq.go           |    20 +
 vendor/src/github.com/coreos/etcd/client/cancelreq_go14.go      |    17 +
 vendor/src/github.com/coreos/etcd/client/client.go              |   514 ++++
 vendor/src/github.com/coreos/etcd/client/cluster_error.go       |    33 +
 vendor/src/github.com/coreos/etcd/client/curl.go                |    70 +
 vendor/src/github.com/coreos/etcd/client/discover.go            |    21 +
 vendor/src/github.com/coreos/etcd/client/doc.go                 |    71 +
 vendor/src/github.com/coreos/etcd/client/keys.generated.go      |   891 +++++++
 vendor/src/github.com/coreos/etcd/client/keys.go                |   644 +++++
 vendor/src/github.com/coreos/etcd/client/members.go             |   272 ++
 vendor/src/github.com/coreos/etcd/client/srv.go                 |    65 +
 vendor/src/github.com/coreos/etcd/pkg/pathutil/path.go          |    29 +
 vendor/src/github.com/coreos/etcd/pkg/types/id.go               |    41 +
 vendor/src/github.com/coreos/etcd/pkg/types/set.go              |   178 ++
 vendor/src/github.com/coreos/etcd/pkg/types/slice.go            |    22 +
 vendor/src/github.com/coreos/etcd/pkg/types/urls.go             |    74 +
 vendor/src/github.com/coreos/etcd/pkg/types/urlsmap.go          |    75 +
 vendor/src/github.com/coreos/go-etcd/LICENSE                    |   202 --
 vendor/src/github.com/coreos/go-etcd/etcd/add_child.go          |    23 -
 vendor/src/github.com/coreos/go-etcd/etcd/client.go             |   481 ----
 vendor/src/github.com/coreos/go-etcd/etcd/cluster.go            |    37 -
 vendor/src/github.com/coreos/go-etcd/etcd/compare_and_delete.go |    34 -
 vendor/src/github.com/coreos/go-etcd/etcd/compare_and_swap.go   |    36 -
 vendor/src/github.com/coreos/go-etcd/etcd/debug.go              |    55 -
 vendor/src/github.com/coreos/go-etcd/etcd/delete.go             |    40 -
 vendor/src/github.com/coreos/go-etcd/etcd/error.go              |    49 -
 vendor/src/github.com/coreos/go-etcd/etcd/get.go                |    32 -
 vendor/src/github.com/coreos/go-etcd/etcd/member.go             |    30 -
 vendor/src/github.com/coreos/go-etcd/etcd/options.go            |    72 -
 vendor/src/github.com/coreos/go-etcd/etcd/requests.go           |   405 ---
 vendor/src/github.com/coreos/go-etcd/etcd/response.go           |    89 -
 vendor/src/github.com/coreos/go-etcd/etcd/set_update_create.go  |   137 -
 vendor/src/github.com/coreos/go-etcd/etcd/version.go            |     6 -
 vendor/src/github.com/coreos/go-etcd/etcd/watch.go              |   103 -
 vendor/src/github.com/docker/libkv/.travis.yml                  |     2 +-
 vendor/src/github.com/docker/libkv/README.md                    |     9 +-
 vendor/src/github.com/docker/libkv/store/consul/consul.go       |    48 +-
 vendor/src/github.com/docker/libkv/store/etcd/etcd.go           |   437 +--
 vendor/src/github.com/docker/libkv/store/store.go               |    12 +-
 vendor/src/github.com/docker/libkv/store/zookeeper/zookeeper.go |    11 +-
 vendor/src/github.com/ugorji/go/LICENSE                         |    22 +
 vendor/src/github.com/ugorji/go/codec/0doc.go                   |   150 ++
 vendor/src/github.com/ugorji/go/codec/README.md                 |   148 ++
 vendor/src/github.com/ugorji/go/codec/binc.go                   |   901 +++++++
 vendor/src/github.com/ugorji/go/codec/cbor.go                   |   566 ++++
 vendor/src/github.com/ugorji/go/codec/decode.go                 |  1552 +++++++++++
 vendor/src/github.com/ugorji/go/codec/encode.go                 |  1232 +++++++++
 vendor/src/github.com/ugorji/go/codec/fast-path.generated.go    | 28307 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 vendor/src/github.com/ugorji/go/codec/fast-path.go.tmpl         |   442 +++
 vendor/src/github.com/ugorji/go/codec/gen-dec-array.go.tmpl     |    80 +
 vendor/src/github.com/ugorji/go/codec/gen-dec-map.go.tmpl       |    46 +
 vendor/src/github.com/ugorji/go/codec/gen-helper.generated.go   |   102 +
 vendor/src/github.com/ugorji/go/codec/gen-helper.go.tmpl        |   250 ++
 vendor/src/github.com/ugorji/go/codec/gen.generated.go          |   139 +
 vendor/src/github.com/ugorji/go/codec/gen.go                    |  1728 ++++++++++++
 vendor/src/github.com/ugorji/go/codec/helper.go                 |   846 ++++++
 vendor/src/github.com/ugorji/go/codec/helper_internal.go        |   151 ++
 vendor/src/github.com/ugorji/go/codec/helper_not_unsafe.go      |    20 +
 vendor/src/github.com/ugorji/go/codec/helper_unsafe.go          |    39 +
 vendor/src/github.com/ugorji/go/codec/json.go                   |   924 +++++++
 vendor/src/github.com/ugorji/go/codec/msgpack.go                |   843 ++++++
 vendor/src/github.com/ugorji/go/codec/noop.go                   |   164 ++
 vendor/src/github.com/ugorji/go/codec/prebuild.go               |     3 +
 vendor/src/github.com/ugorji/go/codec/prebuild.sh               |   193 ++
 vendor/src/github.com/ugorji/go/codec/rpc.go                    |   180 ++
 vendor/src/github.com/ugorji/go/codec/simple.go                 |   505 ++++
 vendor/src/github.com/ugorji/go/codec/test-cbor-goldens.json    |   639 +++++
 vendor/src/github.com/ugorji/go/codec/test.py                   |   119 +
 vendor/src/github.com/ugorji/go/codec/time.go                   |   193 ++
```

Again, if the client library we are currently using is deprecated, I don't contest the move, it's just unfortunate that we have to bring so much in.

dhiltgen commented:
In the interest of reducing risk and trying to get this in for 1.9, I've reworked my TLS change slightly to workaround the bug that the libkv bump solved.  As such, I've removed the libkv vendored bump and etcd dependency change from this PR.  Then we can figure out what to do with libkv+etcd and the deprecated client lib after 1.9.

icecrime commented on daemon/config.go:
```diff
@@ -38,6 +38,7 @@ type CommonConfig struct {
 	// multihost networking (to store networks and endpoints information) and by the node discovery
 	// mechanism.
 	ClusterStore string
+	ClusterOpts  map[string]string
```
Can we please comment the exported fields? We're trying to make this a habit :-)

icecrime commented:
One small nit, otherwise overall LGTM (especially without this hairy dependencies, thanks for this!).

dhiltgen commented:
Added comment explaining ClusterOpts

abronan commented:
Imho we should still get this in because there may be security and/or stability implications (`go-etcd` clearly lists the fact that it's not thread-safe as a caveat) with keeping the old deprecated client :confused:

Having to vendor this is clearly ridiculous but the next `etcd` APIv3 will use `protobuf` instead and we already vendor it so this is going to get rid of this huge dependency..

jfrazelle commented:
wow a lot simplier now <3

vdemeester commented:
So, the idea is to merge it without the vendoring for 1.9, and vendor it in after the release to reduce the risk. 

LGTM :wink:

dhiltgen commented:
@vdemeester correct.  I've worked around the bug for now, so we can stick with the old libkv+etcd lib combo, but as various folks have pointed out, that old etcd lib is deprecated, so we do need to figure out what to do with it before too long (just not for 1.9)

abronan commented:
@vdemeester What risk? :confused: 

This is the same but improved client with a less buggy and supported code base. I really don't see the risk in making the move now, other than having to embed this ridiculous dependency :disappointed: (that will drop for `1.10` anyways)

I will not block this PR to move forward if everyone feels like this is the right call (I know the dependency issue in docker is at its peak), but please remember that we embed a deprecated thing that is marked as unstable in what is supposed to be a stable release of docker.

vdemeester commented:
@abronan I see your point, and I won't block it the other way around :wink:. Was just asking (because I felt the amount of changes dropped a bunch :stuck_out_tongue_closed_eyes:).

dhiltgen commented:
Can I request that the libkv+etcd vendoring bump be switched to a different PR to try to reduce the confusion?  I don't mind submitting it based on the change that was included previously in this PR.

dhiltgen commented:
I've submitted https://github.com/docker/docker/pull/16863 which contains just the libkv+etcd bump independent of my TLS discovery fix.

abronan commented:
@dhiltgen Yes good call! Let's move the discussion in another PR to not block this one.

jfrazelle commented:
LGTM

icecrime commented:
@abronan Just to make sure:

> This is the same but improved client with a less buggy and supported code base. I really don't see the risk in making the move now, other than having to embed this ridiculous dependency :disappointed: (that will drop for 1.10 anyways)

So that means we ship this PR as part of 1.10, and we'll never have this vendoring problem again because we'll directly use the new client package with protobuf (which we already have), correct?

icecrime commented:
Also this merge was a bit rapid, the PR didn't go through docs review :-) @thaJeztah can you please look into it and let us know if something needs to be changed?

jfrazelle commented:
I will make the changes too because it was my bad sorry

dhiltgen commented:
I'm happy to make any doc changes you'd like in a follow up PR :-)

abronan commented:
@icecrime Yes the dependency to `ugorji/go/codec` will most likely drop in `1.10` because the client will use `protobuf` instead (that as you outlined, we already vendor).

@yichengq Sorry for the ping, will the new version of the etcd client supporting APIv3 still embed `ugorji/go/codec`? Right now we discuss about vendoring the new `etcd/coreos/client` but we have issues with the size of this dependency that is 50K lines. Thanks!

thaJeztah commented on docs/reference/commandline/daemon.md:
````diff
@@ -537,6 +538,20 @@ please check the [run](run.md) reference.
+files. For example:
+
+```bash
+    --cluster-advertise 192.168.1.2:2376 \
````
can you add the `docker daemon` command itself here for consistency? i.e.

```bash
$ docker daemon \
    --cluster-advertise 192.168.1.2:2376  \
    ...
```

thaJeztah commented on docs/reference/commandline/daemon.md:
```diff
@@ -537,6 +538,20 @@ please check the [run](run.md) reference.
+the node within the cluster.  Some Key/Value backends support mutual
+TLS, and the client TLS settings used by the daemon can be configured
+using the `--cluster-store-opt` flag, specifying the paths to PEM encoded
+files. For example:
```
Can you add a list with the accepted `kv.*` options?

thaJeztah replied:
The options are fairly straightforward, so the descriptions doesn't have to be long, but for reference; https://github.com/docker/docker/blob/master/docs/reference/commandline/daemon.md#storage-driver-options

thaJeztah commented:
@dhiltgen thanks! the docs changes also need to be added to the man page https://github.com/docker/docker/blob/master/man/docker-daemon.8.md.

It looks like all the discovery flags are undocumented there; 
 `--cluster-store`, `--cluster-advertise` and `--cluster-store-opt`

Also this probably needs an update in the completion scripts, so ping @albers @sdurrheimer (thanks!)

albers commented:
@thaJeztah Done, see #16882

thaJeztah commented:
@albers wow, you're the best, Harald! <3

dhiltgen commented:
I've filed PR https://github.com/docker/docker/pull/16888 to capture the doc requests for this PR.
